tests/socket_test.c

```diff
@@ -1326,6 +1326,232 @@ static int s_cleanup_in_write_cb_doesnt_explode(struct aws_allocator *allocator,
+                break;
+            }
+
+            /* other end must have hung up. clean up and signal completion */
+            aws_socket_clean_up(g_async_tester.read_socket);
+            aws_mem_release(g_async_tester.allocator, g_async_tester.read_socket);
+
+            aws_mutex_lock(g_async_tester.mutex);
+            g_async_tester.read_tasks_complete = true;
+            aws_mutex_unlock(g_async_tester.mutex);
+            aws_condition_variable_notify_all(g_async_tester.condition_variable);
+            break;
+        }
+    }
+}
+
+static void s_async_write_completion(struct aws_socket *socket, int error_code, size_t bytes_written, void *user_data) {
+    enum async_role role = *(enum async_role *)user_data;
+    aws_mem_release(g_async_tester.allocator, user_data);
+
+    /* ensure callback is not firing synchronously from within aws_socket_write() */
+    AWS_FATAL_ASSERT(!g_async_tester.currently_writing);
+
+    /* ensure callbacks arrive in order */
+    AWS_FATAL_ASSERT(g_async_tester.next_expected_callback == role);
+    g_async_tester.next_expected_callback++;
+
+    switch (role) {
+        case ASYNC_ROLE_A_CALLBACK_WRITES_D: {
+            AWS_FATAL_ASSERT(0 == error_code);
+            AWS_FATAL_ASSERT(1 == bytes_written);
+            g_async_tester.currently_writing = true;
+            struct aws_byte_cursor data = aws_byte_cursor_from_c_str("D");
+            enum async_role *d_role = aws_mem_acquire(g_async_tester.allocator, sizeof(enum async_role));
+            *d_role = ASYNC_ROLE_D_GOT_WRITTEN_VIA_CALLBACK;
+            AWS_FATAL_ASSERT(0 == aws_socket_write(socket, &data, s_async_write_completion, d_role));
+            g_async_tester.currently_writing = false;
+            break;
+        }
+        case ASYNC_ROLE_B_CALLBACK_CLEANS_UP_SOCKET:
+            AWS_FATAL_ASSERT(0 == error_code);
+            AWS_FATAL_ASSERT(1 == bytes_written);
+            aws_socket_clean_up(socket);
+            break;
+        case ASYNC_ROLE_C_IS_LAST_FROM_INITIAL_BATCH_OF_WRITES:
+            /* C might succeed or fail (since socket killed after B completes), either is valid */
+            break;
+        case ASYNC_ROLE_D_GOT_WRITTEN_VIA_CALLBACK:
+            /* write tasks complete! */
+            aws_mutex_lock(g_async_tester.mutex);
+            g_async_tester.write_tasks_complete = true;
+            aws_mutex_unlock(g_async_tester.mutex);
+            aws_condition_variable_notify_all(g_async_tester.condition_variable);
+            break;
+        default:
+            AWS_FATAL_ASSERT(0);
+    }
+}
+
+static void s_async_write_task(struct aws_task *task, void *args, enum aws_task_status status) {
+    (void)task;
+    (void)args;
+    (void)status;
+
+    g_async_tester.currently_writing = true;
+
+    struct aws_byte_cursor data = aws_byte_cursor_from_c_str("A");
+    enum async_role *role = aws_mem_acquire(g_async_tester.allocator, sizeof(role));
```

Comment:
I can't figure out how this role is ever freed. What am I missing?

Reply:
https://github.com/awslabs/aws-c-io/pull/353/files#diff-13f9eb7e9121add300ed1f5e22aaf7bc837babd0d535670cf16b4e1628d30a59R1388, it's used as the callback userdata, and cleaned up in the callback.

Reply:
It's freed when the write-completion callback runs